### ISSUE TITLE
NEXT_DYNAMIC_NO_SSR_CODE thrown on not-found page

### ISSUE THREAD

**Issue by altano** [vercel/next.js#53061]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
❯ next info

    Operating System:
      Platform: linux
      Arch: x64
      Version: Ubuntu 20.04.0 LTS Sat Jul 22 2023 19:46:10 GMT-0700 (Pacific Daylight Time)
    Binaries:
      Node: 16.20.0
      npm: 9.4.2
      Yarn: 1.22.19
      pnpm: 8.6.3
    Relevant Packages:
      next: 13.4.12-canary.0
      eslint-config-next: 13.4.12
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.1.6
    Next.js Config:
      output: N/A


warn  - Latest canary version not detected, detected: "13.4.12-canary.0", newest: "13.4.12".
        Please try the latest canary version (`npm install next@canary`) to confirm the issue still exists before creating a new issue.
        Read more - https://nextjs.org/docs/messages/opening-an-issue
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

App Router, Dynamic imports (next/dynamic)

### Link to the code that reproduces this issue or a replay of the bug

https://stackblitz.com/edit/stackblitz-starters-w2wwag?file=src%2Fapp%2Flayout.tsx

### To Reproduce

For setup:
1. You *DO NOT* need turbopack to be enabled (people mentioning this bug in https://github.com/vercel/next.js/issues/50585 thought you needed it, but it's not required)
2. You must dynamically included a client component in the root layout:
```tsx
import dynamic from 'next/dynamic';

const ClientButton = dynamic(() => import('./clientButton'), {
  ssr: false,
});
```

Then:
1. Let the root page load
4. Click the "Click this link to go to the 404 page" link. This just navigates to a page that doesn't exist, rendering not-found.tsx

### Describe the Bug

An error will show up in the terminal and the dev preview error dialog:

```
Unhandled Runtime Error
Error: NEXT_DYNAMIC_NO_SSR_CODE

Call Stack
updateDehydratedSuspenseComponent
node_modules/.pnpm/next@13.4.12-canary.0_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/compiled/react-dom/cjs/react-dom.development.js (16348:0)
```

<img width="1026" alt="image" src="https://github.com/vercel/next.js/assets/1009/82b94004-5810-48ef-b7b7-91caaff62a11">

### Expected Behavior

No errors.

**Comment by john-subba** [comment]
I did get this issue while using next/dynamic with turbo sad to see they haven't fixed it yet since i need both for dev as turbopack is lot faster than normal next dev

**Comment by altano** [comment]
@john-subba the bug does NOT retiree turbo.

**Comment by john-subba** [comment]
> @john-subba the bug does NOT require turbo.

i haven't had this issue without turbo on normal next dev. only got while using turbopack

**Comment by altano** [comment]
In my original post I've attached a reproduction of the issue that does not require turbo.

**Comment by jetaggart** [comment]
We're seeing this only on vercel builds. Local builds fine.

**Comment by huozhi** [comment]
Tested the reproduction it's fixed in the latest version of nextjs (13.4.16)
